### PR TITLE
fix rulenodeserver handling of inputsByTask/task_inputs

### PR DESCRIPTION
**Pull request by barentine**
Addresses issue current ruleserver code doesn't work for making multiple tasks off a single recipe rule.

After fixing bytes/str collision #476, and as mentioned in that PR, you hit a json error 
```
File "/home/ubuntu/miniconda3/lib/python3.6/threading.py", line 864, in run
    self._target(*self._args, **self._kwargs)
  File "/home/ubuntu/code/python-microscopy/PYME/cluster/rulenodeserver.py", line 344, in _poll_tasks
    self._update_tasks()
  File "/home/ubuntu/code/python-microscopy/PYME/cluster/rulenodeserver.py", line 215, in _update_tasks
    taskID, cost = rater.get_local()
  File "/home/ubuntu/code/python-microscopy/PYME/cluster/rulenodeserver.py", line 113, in get_local
    taskID, cost = self.next()
  File "/home/ubuntu/code/python-microscopy/PYME/cluster/rulenodeserver.py", line 107, in next
    return self.__next__()
  File "/home/ubuntu/code/python-microscopy/PYME/cluster/rulenodeserver.py", line 78, in __next__
    task = json.loads(filled_template)
ValueError: Unexpected character in found when decoding object value
```

Even for a nice inputs by task on the ruleserver side (matching the docs there)
```
intputs by task {0: {'input': 'pyme-cluster:///Bergamot/2020_9_28/secret/000/analysis/28_9_series_D.h5r'}, 1: {'input': 'pyme-cluster:///Bergamot/2020_9_28/secret/000/analysis/28_9_series_C.h5r'}, 2: {'input': 'pyme-cluster:///Bergamot/2020_9_28/secret/000/analysis/28_9_series_F.h5r'}, 3: {'input': 'pyme-cluster:///Bergamot/2020_9_28/secret/000/analysis/28_9_series_A.h5r'}, 4: {'input': 'pyme-cluster:///Bergamot/2020_9_28/secret/000/analysis/28_9_series_E.h5r'}, 5: {'input': 'pyme-cluster:///Bergamot/2020_9_28/secret/000/analysis/28_9_series_B.h5r'}}
```
the rulenodeserver template filling fails because it key's out a single input by task (as a dict, which is good) and then for some reason `json.dumps` that and tries to stuff that in as a key in the template. The template is then `json.dumps`ed and then we try to `json.loads` it. This fails, and we end up with something like `{"id": "{{ruleID}}~0", "type": "recipe", "inputs": ""pyme-cluster:///Bergamot/2020_9_28/secret/000/analysis/28_9_series_E.h5r"", `

Anyway, easier to just load the template and then pipe in the task input for `inputs`. This needs to happen in the bidding too.

**Is this a bugfix or an enhancement?**
bugfix
**Proposed changes:**
- don't double-dump load when we can just load once and use the dict

**Checklist:**

- [ ] Tested with numpy=1.14
1.19
- [ ] Tested on python 2.7 and 3.6
3.6
- [ ] Tested with wx=3.x and wx=4.x [if UI code]
- [x] Does the PR avoid variable renaming in existing code, whitespace changes, and other forms of tidying? [There is a place for code tidying, but it makes reviewing 
much simpler if this is kept separate from functional changes]

### CODE REVIEW

**Comment by barentine** [comment]
@David-Baddeley if you can have a look at this so we can discuss tomorrow - it's separate of HTTPRulePusher and _rules.RecipeRule so I was hoping we might be able to shoe it in easy

File: PYME/cluster/rulenodeserver.py [hunk 1]
```diff
@@ -35,7 +35,7 @@ def template_fill(template, **kwargs):
     s = template
 
     for key, value in kwargs.items():
-        s = s.replace('{{%s}}' % key, '%s' % value)
+        s = s.replace('{{%s}}' % key, str(value))
```
Comment (David-Baddeley) [inline PYME/cluster/rulenodeserver.py:38]: probably doesn't matter either way, but curious as to whether this change was needed. Did the original trigger an error?

**Comment by David-Baddeley** [comment]
Will probably refactor so the task completion logic is all in one function rather than being duplicated in rater and task queueing rather than merging directly. As it is it looks like we are modifying the same thing in two places.

**Comment by David-Baddeley** [comment]
I'm a little uncomfortable with the concept of over-writing the 'inputs' entry in the task once we create it from the template (conceptually the task template should specify what the task looks like, modulo any templated substitutions - over-writing the task inputs introduces "magic" that I'd like to avoid - we'd certainly need to change the docs, but I'd prefer to fix the templating if possible.

What really puzzles me is why the existing code doesn't work:

```python
>>> from PYME.cluster import rulenodeserver
>>> import json
>>> rule_template = '''{
         "id" : "{{ruleID}}~{{taskID}}",
         "type" : "recipe",
         "taskdef" : {"recipe" : "<recipe as a YAML string>"},
         "inputs" : {{taskInputs}},
         "output_dir" : "PYME-CLUSTER://path/to/output/dir" 
    }'''
>>> filled_template = rulenodeserver.template_fill(rule_template, ruleID=0, taskID=0, taskInputs=json.dumps({'input' : 'foo/bar.txt'})
>>> print(filled_template)
{
         "id" : "0~0",
         "type" : "recipe",
         "taskdef" : {"recipe" : "<recipe as a YAML string>"},
         "inputs" : '{"input": "foo/bar.txt"}',
         "output_dir" : "PYME-CLUSTER://path/to/output/dir"
}
>>> json.loads(filled_template)
{'id': '0~0',
 'inputs': {'input': 'foo/bar.txt'},
 'output_dir': 'PYME-CLUSTER://path/to/output/dir',
 'taskdef': {'recipe': '<recipe as a YAML string>'},
 'type': 'recipe'}
```

**Comment by David-Baddeley** [comment]
aside from the `strict=False` parameter to `json.loads()` described in the bug report for multi-line text (which this PR doesn't address)

**Comment by David-Baddeley** [comment]
Actually, now wondering if this might be a `ujson` version issue (everything works fine for me on py=3.6, ujson=1.35, even without `strict=False`)

File: PYME/cluster/rulenodeserver.py (hunk 1 above)
Comment (barentine) [reply PYME/cluster/rulenodeserver.py:38]: not needed, but after we ran into a weird issues with #403 /  #407 I thought it was worthwhile while we were here

**Comment by barentine** [comment]
I'm using ujson version 3.0.0 and 3.1.0. 

Your example works because your yaml str isn't a multiline yaml str. 

```
In [1]: from PYME.cluster import rulenodeserver                                 

In [2]: import json                                                             

In [3]: rule_template = '''{ 
   ...:          "id" : "{{ruleID}}~{{taskID}}", 
   ...:          "type" : "recipe", 
   ...:          "taskdef" : {"recipe" : "<recipe as a YAML string>"}, 
   ...:          "inputs" : {{taskInputs}}, 
   ...:          "output_dir" : "PYME-CLUSTER://path/to/output/dir"  
   ...:     }'''                                                                

In [4]: filled_template = rulenodeserver.template_fill(rule_template, ruleID=0, 
   ...: taskID=0, taskInputs=json.dumps({'input' : 'foo/bar.txt'}))             

In [5]: print(filled_template)                                                  
{
         "id" : "0~0",
         "type" : "recipe",
         "taskdef" : {"recipe" : "<recipe as a YAML string>"},
         "inputs" : {"input": "foo/bar.txt"},
         "output_dir" : "PYME-CLUSTER://path/to/output/dir" 
    }

In [6]: json.loads(filled_template)                                             
Out[6]: 
{'id': '0~0',
 'type': 'recipe',
 'taskdef': {'recipe': '<recipe as a YAML string>'},
 'inputs': {'input': 'foo/bar.txt'},
 'output_dir': 'PYME-CLUSTER://path/to/output/dir'}

In [7]: rule_template = '''{ 
   ...:          "id" : "{{ruleID}}~{{taskID}}", 
   ...:          "type" : "recipe", 
   ...:          "taskdef" : {"recipe" : "<recipe as a YAML string> \r\n some st
   ...: uff"}, 
   ...:          "inputs" : {{taskInputs}}, 
   ...:          "output_dir" : "PYME-CLUSTER://path/to/output/dir"  
   ...:     }'''                                                                

In [8]: filled_template = rulenodeserver.template_fill(rule_template, ruleID=0, 
   ...: taskID=0, taskInputs=json.dumps({'input' : 'foo/bar.txt'}))             

In [9]: json.loads(filled_template)                                             
---------------------------------------------------------------------------
JSONDecodeError                           Traceback (most recent call last)
<ipython-input-9-a8fe9fc0cfb6> in <module>
----> 1 json.loads(filled_template)

~/miniconda3/lib/python3.6/json/__init__.py in loads(s, encoding, cls, object_hook, parse_float, parse_int, parse_constant, object_pairs_hook, **kw)
    352             parse_int is None and parse_float is None and
    353             parse_constant is None and object_pairs_hook is None and not kw):
--> 354         return _default_decoder.decode(s)
    355     if cls is None:
    356         cls = JSONDecoder

~/miniconda3/lib/python3.6/json/decoder.py in decode(self, s, _w)
    337 
    338         """
--> 339         obj, end = self.raw_decode(s, idx=_w(s, 0).end())
    340         end = _w(s, end).end()
    341         if end != len(s):

~/miniconda3/lib/python3.6/json/decoder.py in raw_decode(self, s, idx)
    353         """
    354         try:
--> 355             obj, end = self.scan_once(s, idx)
    356         except StopIteration as err:
    357             raise JSONDecodeError("Expecting value", s, err.value) from None

JSONDecodeError: Invalid control character at: line 4 column 61 (char 113)

```

further, your example works because you didn't use a dict `inputsByTask` (dict of dicts) to fill `task_inputs`, which is what this PR addresses:

```
In [10]: filled_template = rulenodeserver.template_fill(rule_template, ruleID=0,
    ...:  taskID=0, taskInputs=json.dumps({0: {'input': 'pyme-cluster:///Bergamo
    ...: t/2020_9_28/secret/000/analysis/28_9_series_D.h5r'}, 1: {'input': 'pyme
    ...: -cluster:///Bergamot/2020_9_28/secret/000/analysis/28_9_series_C.h5r'},
    ...:  2: {'input': 'pyme-cluster:///Bergamot/2020_9_28/secret/000/analysis/2
    ...: 8_9_series_F.h5r'}, 3: {'input': 'pyme-cluster:///Bergamot/2020_9_28/se
    ...: cret/000/analysis/28_9_series_A.h5r'}, 4: {'input': 'pyme-cluster:///Be
    ...: rgamot/2020_9_28/secret/000/analysis/28_9_series_E.h5r'}, 5: {'input': 
    ...: 'pyme-cluster:///Bergamot/2020_9_28/secret/000/analysis/28_9_series_B.h
    ...: 5r'}}))                                                                

In [11]: json.loads(filled_template)                                            
---------------------------------------------------------------------------
JSONDecodeError                           Traceback (most recent call last)
<ipython-input-11-a8fe9fc0cfb6> in <module>
----> 1 json.loads(filled_template)

~/miniconda3/lib/python3.6/json/__init__.py in loads(s, encoding, cls, object_hook, parse_float, parse_int, parse_constant, object_pairs_hook, **kw)
    352             parse_int is None and parse_float is None and
    353             parse_constant is None and object_pairs_hook is None and not kw):
--> 354         return _default_decoder.decode(s)
    355     if cls is None:
    356         cls = JSONDecoder

~/miniconda3/lib/python3.6/json/decoder.py in decode(self, s, _w)
    337 
    338         """
--> 339         obj, end = self.raw_decode(s, idx=_w(s, 0).end())
    340         end = _w(s, end).end()
    341         if end != len(s):

~/miniconda3/lib/python3.6/json/decoder.py in raw_decode(self, s, idx)
    353         """
    354         try:
--> 355             obj, end = self.scan_once(s, idx)
    356         except StopIteration as err:
    357             raise JSONDecodeError("Expecting value", s, err.value) from None

JSONDecodeError: Invalid control character at: line 4 column 61 (char 113)

```

There is no reason to json dumps and undumps taskInputs, when we could instead just pop it straight into the task dict

**Comment by David-Baddeley** [comment]
As mentioned above, I did test a multiline recipe (just didn't include all of code) - it works without an error if you `import ujson as json` (as we do in the ruleserver) and needs `strict=False` if you use the stock json module, but is in some ways a separate issue to the changes in the PR. I also skipped a few steps that I thought were obvious above (e.g. getting the specific task  inputs from `inputsByTask`), but this works on my py3 ... 

```python
>>> from PYME.cluster import rulenodeserver
>>> import ujson as json
>>> rule_template = '''{
         "id" : "{{ruleID}}~{{taskID}}",
         "type" : "recipe",
         "taskdef" : {"recipe" : "<recipe as a YAML string\n with multiple lines>"},
         "inputs" : {{taskInputs}},
         "output_dir" : "PYME-CLUSTER://path/to/output/dir" 
    }'''
>>> inputsByTask = {0:{'input': 'foo/bar.txt'}, 1:{'input':'whiz/bang.txt'}}
>>> taskID = 0
>>> filled_template = rulenodeserver.template_fill(rule_template, ruleID=0, taskID=taskID,
                                                   taskInputs=json.dumps(inputsByTask.get(taskID)))
>>> json.loads(filled_template)
{'id': '0~0',
 'inputs': {'input': 'foo/bar.txt'},
 'output_dir': 'PYME-CLUSTER://path/to/output/dir',
 'taskdef': {'recipe': '<recipe as a YAML string>'},
 'type': 'recipe'}
```

One important clarification is that by the time you are setting taskInputs, you should have got the specific inputs for that task (i.e. it should no longer be a dict of dict).

My reluctance about directly overwriting  `inputs` in the task itself is explained as follows: 
- with the rule-server we make an implicit contract that we will only change a task by string substitution. IE your task is what you submit except where you have specifically flagged it to change/be substituted
- if we start over-writing `inputs` outside the templating we have to think about what we should do, e.g. when someone submits a rule which does has a standard (non-templated) `inputs` section in the rule template and also provides `inputsByTask` in the rule (this should ideally not happen, but currently has predictable behaviour with `inputsByTask` ignored as there is no-where in the template to substitute into).

I accept that it's largely a semantic issue, and would likely be OK with the hacky work-around (subject to changing the ruleserver docstrings to make it clear that inputs might be over-written) if it was clear to me a) that it actually fixed the problem and  b) why it fixed it / what exactly is broken. As it stands, I can't make the current code fail (at least when broken down into the important bits), so I suspect that the actual bug is elsewhere. This feels like change for the sake of change which makes the code slightly less predicatable and probably doesn't address the underlying issue.

**Comment by barentine** [comment]
Fair enough re: not wanting to make changes unless it's very clear as to why. I obviously resent the sentiment that I am pushing for 'change[s] for the sake of change', but will once more see if I can run a recipe from the templated clusterUI view and report back.

**Comment by barentine** [comment]
all after git pull, git reset --hard HEAD, running a recipe from `Selected Files > Run Recipe` testing a potential fix (grab 'series' and 'files' rather than just 'files' for recipe view) to #539 
image

ujson versions 3.0.0 and 3.1.0

**Comment by barentine** [comment]
> One important clarification is that by the time you are setting taskInputs, you should have got the specific inputs for that task (i.e. it should no longer be a dict of dict).

What do you mean by this, David? `HTTPRulePusher.RecipePusher` forces `inputsByTask` as a dict of dict. 

```
        inputs_by_task = {frameNum: {k : inputs[k][frameNum] for k in inputs.keys()} for frameNum in range(numTotalFrames)}

        rule = {'template': self._taskTemplate, 'inputsByTask' : inputs_by_task}
        print('intputs by task %s' % inputs_by_task)
        s = clusterIO._getSession(self.taskQueueURI)
        r = s.post('%s/add_integer_id_rule?max_tasks=%d&release_start=%d&release_end=%d' % (self.taskQueueURI,numTotalFrames, 0, numTotalFrames), data=json.dumps(rule),
                        headers = {'Content-Type': 'application/json'})
```

```
intputs by task {0: {'input': 'pyme-cluster://villains/Bergamot/2020_10_22/000/reg0.pcs'}}
```

**Comment by barentine** [comment]
Possibly tangential, but ujson does not have the strict flag available.
```
In [1]: import ujson                                                                                                                     

In [2]: ujson.__version__                                                                                                                
Out[2]: '1.35'

In [3]: ujson.loads(ujson.dumps({'test': 'ok'}))                                                                                         
Out[3]: {'test': 'ok'}

In [4]: ujson.loads(ujson.dumps({'test': 'ok'}), strict=False)                                                                           
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-4-6dfb5646c5d1> in <module>
----> 1 ujson.loads(ujson.dumps({'test': 'ok'}), strict=False)

TypeError: 'strict' is an invalid keyword argument for this function

In [5]: import json                                                                                                                      

In [6]: json.loads(json.dumps({'test': 'ok'}), strict=False)                                                                             
Out[6]: {'test': 'ok'}

```

**Comment by barentine** [comment]
Wanted to follow up after closing #540 - I still hit the same error on the nodeserver. This time using the recipe form (selected files) and the following (mundane) recipe
```
- tablefilters.FilterTable:
    filters:
      LLH:
      - -250.0
      - 0.0
      error_x:
      - 0.0
      - 20.0
      error_y:
      - 0.0
      - 20.0
    inputName: fit_results
    outputName: filtered
- measurement.DetectPoints2D:
    input_name: input
    output_name: candidate_points
- measurement.FitPoints:
    fitModule: AstigGaussGPUFitFR
    inputImage: input
    inputPositions: candidate_points
    outputName: fit_results
- multiview.Fold:
    input_name: filtered
    output_name: folded
- multiview.CalibrateShifts:
    input_name: folded
    output_name: shiftmap
    search_radius_nm: 300.0
- output.HDFOutput:
    filePattern: '{output_dir}/{file_stub}.sf'
    inputVariables:
      shiftmap: shift_map
    scheme: pyme-cluster://
```

**Comment by barentine** [comment]
The new `rules.RecipeRule` does not set `inputs` in the template, nor offer an interface to. Anything you pass to initialize (e.g. #486) will get you a filled `inputsByTask` and a `inputs: {{taskInputs}}` - printing `repr(self.rule)` in `Rule._post_rule`:
```
{'template': '{"id": "{{ruleID}}~{{taskID}}",\n                      "type": "recipe",\n                      "inputs" : {{taskInputs}},\n                      "taskdefRef" : "pyme-cluster://villains/RECIPES/calibrate-multiview-shifts.yaml",\n                      "output_dir": "pyme-cluster:///Bergamot/2020_10_28/CalibrationStacks_real/000/analysis/analysis",\n                      }', 'inputsByTask': {0: {'input': 'pyme-cluster:///Bergamot/2020_10_28/CalibrationStacks_real/000/analysis/28_10_series_L_reg.h5r'}}}
```
which itself is fine, but as we've been discussing, on later ujson versions this gets screwed up by the rulenodeserver dumps/loads.

```
(base) smeagol@smeagol:~$ ipython
Python 3.6.9 |Anaconda, Inc.| (default, Jul 30 2019, 19:07:31) 
Type 'copyright', 'credits' or 'license' for more information
IPython 7.16.1 -- An enhanced Interactive Python. Type '?' for help.

In [1]: import ujson as json

In [2]: def template_fill(template, **kwargs):
   ...:     s = template
   ...: 
   ...:     for key, value in kwargs.items():
   ...:         s = s.replace('{{%s}}' % key, '%s' % value)
   ...: 
   ...:     return s
   ...: 

In [3]: rule = {'template': '{"id": "{{ruleID}}~{{taskID}}",\n                  
   ...:     "type": "recipe",\n                      "inputs" : {{taskInputs}},\
   ...: n                      "taskdefRef" : "pyme-cluster://villains/RECIPES/c
   ...: alibrate-multiview-shifts.yaml",\n                      "output_dir": "p
   ...: yme-cluster:///Bergamot/2020_10_28/CalibrationStacks_real/000/analysis/a
   ...: nalysis",\n                      }', 'inputsByTask': {0: {'input': 'pyme
   ...: -cluster:///Bergamot/2020_10_28/CalibrationStacks_real/000/analysis/28_1
   ...: 0_series_L_reg.h5r'}}}

In [4]: inputs = rule.get('inputsByTask', {})

In [5]: inputs
Out[5]: {0: {'input': 'pyme-cluster:///Bergamot/2020_10_28/CalibrationStacks_real/000/analysis/28_10_series_L_reg.h5r'}}

In [6]: inputs = {int(k):v for k, v in inputs.items()}

In [7]: inputs
Out[7]: {0: {'input': 'pyme-cluster:///Bergamot/2020_10_28/CalibrationStacks_real/000/analysis/28_10_series_L_reg.h5r'}}

In [8]: task_inputs = inputs.get(taskID)
---------------------------------------------------------------------------
NameError                                 Traceback (most recent call last)
<ipython-input-8-38c669b91e1a> in <module>
----> 1 task_inputs = inputs.get(taskID)

NameError: name 'taskID' is not defined

In [9]: task_inputs = inputs.get(0)

In [10]: task_inputs
Out[10]: {'input': 'pyme-cluster:///Bergamot/2020_10_28/CalibrationStacks_real/000/analysis/28_10_series_L_reg.h5r'}

In [11]: if not task_inputs is None:
    ...:     task_inputs = json.dumps(task_inputs)
    ...: 

In [12]: task_inputs
Out[12]: '{"input":"pyme-cluster:\\/\\/\\/Bergamot\\/2020_10_28\\/CalibrationStacks_real\\/000\\/analysis\\/28_10_series_L_reg.h5r"}'

In [13]: filled_template = template_fill(rule['template'], taskID=0, taskInputs=
    ...: task_inputs)

In [14]: task = json.loads(filled_template)
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-14-87856a90e264> in <module>
----> 1 task = json.loads(filled_template)

ValueError: Unexpected character in found when decoding object value

In [15]: filled_template
Out[15]: '{"id": "{{ruleID}}~0",\n                      "type": "recipe",\n                      "inputs" : {"input":"pyme-cluster:\\/\\/\\/Bergamot\\/2020_10_28\\/CalibrationStacks_real\\/000\\/analysis\\/28_10_series_L_reg.h5r"},\n                      "taskdefRef" : "pyme-cluster://villains/RECIPES/calibrate-multiview-shifts.yaml",\n                      "output_dir": "pyme-cluster:///Bergamot/2020_10_28/CalibrationStacks_real/000/analysis/analysis",\n                      }'

In [16]: print(repr(filled_template))
'{"id": "{{ruleID}}~0",\n                      "type": "recipe",\n                      "inputs" : {"input":"pyme-cluster:\\/\\/\\/Bergamot\\/2020_10_28\\/CalibrationStacks_real\\/000\\/analysis\\/28_10_series_L_reg.h5r"},\n                      "taskdefRef" : "pyme-cluster://villains/RECIPES/calibrate-multiview-shifts.yaml",\n                      "output_dir": "pyme-cluster:///Bergamot/2020_10_28/CalibrationStacks_real/000/analysis/analysis",\n                      }'

In [17]: ujson.__version__
---------------------------------------------------------------------------
NameError                                 Traceback (most recent call last)
<ipython-input-17-f78e0e8172e7> in <module>
----> 1 ujson.__version__

NameError: name 'ujson' is not defined

In [18]: json.__version__
Out[18]: '4.0.1'

```

What I'm hearing is that if I preserve the behavior of `inputsByTask` being ignored if `{{taskInputs}}` isn't in the template, we can move ahead with this.

**Comment by barentine** [comment]
Following up:
I've changed this PR so we don't clobber `inputs` if it's already set. Additionally the logic for that is refactored, as suggested, and if I have this PR branch merged with #561 (and #486), I finally have the chance to learn 

(drum-roll please)

I forgot to update my recipe after #284 :upside_down_face: 
But I think we're close to rules ruling the world _and_ having clusterUI recipes working again!

```
{"id": "{{ruleID}}~0", "type": "recipe", "inputs": "{{taskInputs}}", "output_dir": "pyme-cluster:///Bergamot/2020_10_28/CalibrationStacks_real/000/analysis/analysis", "taskdefRef": "pyme-cluster://villains/RECIPES/calibrate-multiview-shifts.yaml"}
{"id": "{{ruleID}}~0", "type": "recipe", "inputs": "task_inputs", "output_dir": "pyme-cluster:///Bergamot/2020_10_28/CalibrationStacks_real/000/analysis/analysis", "taskdefRef": "pyme-cluster://villains/RECIPES/calibrate-multiview-shifts.yaml"}
{"id": "000000-5267959c11fc469b850b16a4c4e56ffb~0", "type": "recipe", "inputs": "{{taskInputs}}", "output_dir": "pyme-cluster:///Bergamot/2020_10_28/CalibrationStacks_real/000/analysis/analysis", "taskdefRef": "pyme-cluster://villains/RECIPES/calibrate-multiview-shifts.yaml"}
{"id": "000000-5267959c11fc469b850b16a4c4e56ffb~0", "type": "recipe", "inputs": "task_inputs", "output_dir": "pyme-cluster:///Bergamot/2020_10_28/CalibrationStacks_real/000/analysis/analysis", "taskdefRef": "pyme-cluster://villains/RECIPES/calibrate-multiview-shifts.yaml"}
DEBUG:nodeserver:Launching worker processors
Trying to load 3rd party recipe module bleaching_tools.python_microscopy.recipe_modules.masked_statistics
Trying to load 3rd party recipe module nep_fitting.recipe_modules.nep_fits
Loaded 3rd party recipe module nep_fitting.recipe_modules.nep_fits
Trying to load 3rd party recipe module nucleus.python_microscopy.recipe_modules.distance_to_membrane
Trying to load 3rd party recipe module nucleus.python_microscopy.recipe_modules.shell_fitting
Trying to load 3rd party recipe module quant_condensate.recipe_modules.cluster_quant
Loaded 3rd party recipe module quant_condensate.recipe_modules.cluster_quant
Trying to load 3rd party recipe module nucleus.python_microscopy.recipe_modules.astigmatism_tools
Trying to load 3rd party recipe module cb_analysis.python_microscopy.recipe_modules.cb_clustering
Loaded 3rd party recipe module cb_analysis.python_microscopy.recipe_modules.cb_clustering
Trying to load 3rd party recipe module nucleus.python_microscopy.recipe_modules.nucleus_metrics
Trying to load 3rd party recipe module pyme_omero.recipe_modules.omero_upload
Loaded 3rd party recipe module pyme_omero.recipe_modules.omero_upload
Trying to load 3rd party recipe module cell_death_detection.python_microscopy.recipe_modules.cell_death
Trying to load 3rd party recipe module localization_survey.recipe_modules.roi_cropping
Loaded 3rd party recipe module localization_survey.recipe_modules.roi_cropping
Trying to load 3rd party recipe module pyme_rcnn.recipe_modules.rcnn_segment
Loaded 3rd party recipe module pyme_rcnn.recipe_modules.rcnn_segment
Traceback (most recent call last):
  File "/home/smeagol/code/python-microscopy/PYME/cluster/taskWorkerHTTP.py", line 340, in computeLoop
    recipe.execute()
  File "/home/smeagol/code/python-microscopy/PYME/recipes/base.py", line 640, in execute
    m.check_inputs(self.namespace)
  File "/home/smeagol/code/python-microscopy/PYME/recipes/base.py", line 137, in check_inputs
    raise MissingInputError('Input "%s" is missing from namespace; keys: %s' % (input, keys))
PYME.recipes.base.MissingInputError: Input "input" is missing from namespace; keys: ['FitResults']
Traceback (most recent call last):
  File "/home/smeagol/code/python-microscopy/PYME/IO/h5rFile.py", line 247, in _pollQueues
    self._h5file.flush()
  File "/home/smeagol/miniconda3/lib/python3.6/site-packages/tables/file.py", line 2696, in flush
    self._check_open()
  File "/home/smeagol/miniconda3/lib/python3.6/site-packages/tables/file.py", line 2150, in _check_open
    raise ClosedFileError("the file object is closed")
tables.exceptions.ClosedFileError: the file object is closed
```

**Comment by David-Baddeley** [comment]
Just double checked on the most recent ujson version. It will load strings with new-lines just fine, and has no issues with the `dumps`ed inputs, so this is not a ujson version issue. The error on your example above comes from an extra comma in the task template (which might be addressed in #561), and is independent of inputs. I suspect that this will be un-necessary/redundant once we've sorted out the commas etc ...

**Comment by barentine** [comment]
definitely addressed in #561. 

Can't remember if I tested #561 without this branch merged. If I recall there are two different json errors I was bumping into - one is the `ValueError: Unexpected character in found when decoding object value` and the other has `true` in it, but I'm not finding it in my notes at the moment and I'm probably spun around a bit on the whole thing.

**Comment by barentine** [comment]
Just tested. Still need this in order not to get 
```
File "/home/ubuntu/code/python-microscopy/PYME/cluster/rulenodeserver.py", line 215, in _update_tasks
    taskID, cost = rater.get_local()
  File "/home/ubuntu/code/python-microscopy/PYME/cluster/rulenodeserver.py", line 113, in get_local
    taskID, cost = self.next()
  File "/home/ubuntu/code/python-microscopy/PYME/cluster/rulenodeserver.py", line 107, in next
    return self.__next__()
  File "/home/ubuntu/code/python-microscopy/PYME/cluster/rulenodeserver.py", line 78, in __next__
    task = json.loads(filled_template)
ValueError: Unexpected character in found when decoding object value
```

When running 
```
- measurement.DetectPoints2D:
    input_name: input
    output_name: candidate_points
- measurement.FitPoints:
    fitModule: AstigGaussGPUFitFR
    inputImage: input
    inputPositions: candidate_points
    outputName: fit_results
- tablefilters.FilterTable:
    filters:
      LLH:
      - -250.0
      - 0.0
      error_x:
      - 0.0
      - 20.0
      error_y:
      - 0.0
      - 20.0
    inputName: fit_results
    outputName: filtered
- multiview.Fold:
    input_name: filtered
    output_name: folded
- multiview.CalibrateShifts:
    input_name: folded
    output_name: shiftmap
    search_radius_nm: 300.0
- output.HDFOutput:
    filePattern: '{output_dir}/{file_stub}.sf'
    inputVariables:
      shiftmap: shift_map
    scheme: pyme-cluster://
```

from the clusterUI run view

**Comment by barentine** [comment]
superceded

**Pull request closed without merging**